Support tuple inputs in NN Module test & add parity test for Transformer

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44853 Support tuple inputs in NN Module test & add parity test for Transformer**

This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer, also test MultiHeadAttention as well.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I enabled gradgrad check for TransformerEncoderLayer_relu test only since it is too slow. (10+ seconds for TransformerEncoderLayer, 50+ seconds for TransformerDecoderLayer, 10+ mins for multilayer Transformer). The underlying for these modules are all the same (Dropout, Linear and LayerNorm), so enabled 1 gradgrad test should be enough.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

Commit:
Support tuple inputs in NN Module test

[ghstack-poisoned]

Commit:
Update on "Support tuple inputs in NN Module test"


Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]

glaringlee commented:
This code change is tested by test_nn.py script that is running in multiple CI tests.

Commit:
Update on "Support tuple inputs in NN Module test & add parity test for Transformer"


This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I disabled gradgrad check for Transformer since it is too slow. (20+ mins for simple transformer module). But they are enabled for its underlying TransformerEncoderLayer and TransformerDecoderLayer, so we still covered all the needed tests.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3488,6 +3488,62 @@ def fractional_max_pool3d_test(test_case):
+        check_gradgrad=False,
+        desc='gelu_activation',
+    ),
+    dict(
+        module_name='Transformer',
+        constructor_args=(4, 2, 2, 2, 8, 0.0, "relu"),
+        cpp_constructor_args='''torch::nn::TransformerOptions()
+                                .d_model(4)
+                                .nhead(2)
+                                .num_encoder_layers(2)
+                                .num_decoder_layers(2)
+                                .dim_feedforward(8)
+                                .dropout(0.0)
+                                .activation(torch::kReLU)''',
+        input_fn=lambda:(torch.rand(3, 3, 4), torch.rand(2, 3, 4), torch.rand(3, 3)),
+        check_gradgrad=False,
+        desc='multilayer_coder'
+    )
```
Reading the documentation for Transformer, I would have expected more rigorous testing. The forward function of Transformer takes in up to 8 arguments; we've only tested three here.

Furthermore, for some of those arguments, like `src_mask`, different behavior occurs whether or not it is a ByteTensor / BoolTensor / FloatTensor. Without knowing anything about the underlying Transformer implementation, I think it would be nice to add a variety of test cases to cover the various uses of Transformer

zou3519 replied:
Hmm, I see the problem though: we assume all of the inputs are differentiable and then somewhere down the line, gradients get tested for those inputs

glaringlee replied:
Yes, we treat all the inputs as differentiable. And test jacobian, grad, and even gradgrad.
Testing 3 args vs 8 args have the same effort for parity test, but save time.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4905,120 +4989,124 @@ def _do_test(self, test_case, module, input):
 
             if torch.cuda.device_count() > 1:
-                input = input.cuda(1)
+                input_tuple = tuple(t.cuda(1) for t in input_tuple)
```
For PyTorch C++ API Parity -- are you planning on adding a Parity test for nn.LSTM at some point?

This PR extends NewModuleTest to accept tuples of floating-point tensors as input. However, supporting something like nn.LSTM would require supporting more than just tuples of floating-point tensors. LSTM takes in (input, (hidden_x, cell_x)) arguments. It might be worth sitting down and thinking about how to generalize NewModuleTest if we want to check parity for nn.LSTM in the future.

glaringlee replied:
Yes, this will be done after adding `input_dict`

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3488,6 +3488,62 @@ def fractional_max_pool3d_test(test_case):
+                                .dim_feedforward(8)
+                                .dropout(0.0)
+                                .activation(torch::kReLU)''',
+        input_fn=lambda:(torch.rand(3, 3, 4), torch.rand(2, 3, 4), torch.rand(3, 3)),
```
(thinking out loud) We could try to add more structure to improve the testing here. Maybe we can create a "ModuleInputs" class and say that input_fn must return either a single Tensor (legacy) or a ModuleInputs class.

ModuleInputs would hold a list of inputs and would maintain a difference between optimizable inputs and non-optimizable inputs so that it's possible to provide non-differentiable inputs and the testing harness can avoid computing gradients for those inputs.

glaringlee replied:
Yep. This is something I plan to do, but need bigger amount of work. Adding an `input_dict` which will use number as keys (to keep the order of parameter), and value will be another dict, this dict contains the input value, and additional info. (eg. whether it is differentiable, whether it is a target) So the NewModuleTest and parity test script will be adjusted to handle the `input_dict` correctly.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4566,7 +4622,8 @@ def check_jacobian(self, module, input, jacobian_input=True):
 
         # TODO: compare structure
-        if input.numel() != 0:
+        input_tuple = input if isinstance(input, tuple) else (input,)
+        if any(t.numel() != 0 for t in input_tuple):
```
It looks like the `input` to `check_jacobian` is *always* the tupled inputs. i.e., on line 4953 we have `test_case.check_jacobian(module, input_tuple, self.jacobian_input)`, which looks like the only callsite of check_jacobian.

Is that true? If so, we should update the `check_jacobian` function to take in `self, module, inputs_tuple` and remove this conditional.

glaringlee replied:
This is something I am not sure about. Before this change, check_jacobian accept both individual tensor and tuple of tensors. After this change, we no longer have individual tensor being passed in, but should we remove the support of individual tensor for check_jacobian? I prefer keep the support rather than removing, what do you think?

zou3519 replied:
I find it confusing when `input` can be "either a Tensor or a Tuple of Tensors", especially if our testing code never actually exercises the "Tensor" path. I'm fine if you'd like to leave it in -- in that case, would prefer some type annotations so that a code reader has some idea of what is going on: https://github.com/pytorch/pytorch/pull/44853#discussion_r492341897

glaringlee replied:
I will add type-annotate. I gained a lot of benefit from downstream functions (eg. to_gpu())  which supported tuple tensor already, that made my change much easier. So I think it would be better to keep what we already supported.

zou3519 commented on test/test_nn.py:
```diff
@@ -314,13 +314,17 @@ class TestNN(NNTestCase):
         with freeze_rng_state():
-            return module(input)
+            if isinstance(input, tuple):
+                return module(*input)
+            else:
+                return module(input)
 
     def _backward(self, module, input, output, grad_output, create_graph=False):
         output.backward(grad_output, retain_graph=True, create_graph=create_graph)
-        if input.grad is None:
-            return None
-        return input.grad.data
+        if isinstance(input, tuple):
+            return tuple(map(lambda i: i.grad.data if i.grad is not None else None, input))
+        else:
+            return input.grad.data if input.grad is not None else None
```
Do you think it would be a better API if we just change _forward and _backward such that `input` must always be a tuple? The way it is right now (input can be either Tensor or Tuple) I'm worried that folks will get confused; in addition, the NN module test code is already pretty long and confusing

zou3519 replied:
I'm OK with _forward and _backward being able to take `input` as either a Tensor or as a tuple of inputs if we properly type-annotate it with https://github.com/pytorch/pytorch/blob/9dc2bcdc070cd05c2755c6d9b3e9d74a0b1c306c/torch/types.py#L9

glaringlee replied:
I have the same feeling about this as check_jacobian one. I think we should keep support both, individual tensor and tuple of tensors. I will update the type-annotate to the `input`.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4889,12 +4949,36 @@ def __init__(self, *args, **kwargs):
+        test_case.check_jacobian(module, input_tuple, self.jacobian_input)
         if self.check_gradgrad:
+            # this function creates a lambda function based on the number of tensors in the input
+            # eg. if input is (tensor1, tensor2), the generated lambda function will be:
+            #     lambda _0, _1, *args, **kw: test_case._forward(module, (_0, _1, ))
+            def get_lambda_func(inputs, test_case, module):
+                # we assume the inputs is always not empty
+                len_inputs = len(inputs)
+                if len_inputs == 1:
+                    return lambda _0, *args, **kw: test_case._forward(module, (_0, ))
+                if len_inputs == 2:
+                    return lambda _0, _1, *args, **kw: test_case._forward(module, (_0, _1, ))
+                # generate lambda function for case that length of input > 2
+                helper_str = "def _get_lambda_func(test_case, module):"
+                fn_str = "return lambda "
+                in_str = ""
+                for i in range(0, len(inputs)):
+                    in_str = in_str + "_" + str(i) + ", "
+                fn_str = fn_str + in_str + "*args, **kw: test_case._forward(module, (" + in_str + "))"
+                helper_str = helper_str + fn_str
+                # we have to do exec to define a function in order to pass test_case and module in
+                tmp_locals = {}
+                exec(helper_str, {}, tmp_locals)
+                return tmp_locals['_get_lambda_func'](test_case, module)
+
             # could probably unify check_jacobian above with this.
             params = tuple(x for x in module.parameters())
-            _assertGradAndGradgradChecks(test_case,
-                                         lambda x, *args, **kw: test_case._forward(module, x), (input,) + params)
+            _assertGradAndGradgradChecks(
+                test_case, get_lambda_func(input_tuple, test_case, module), input_tuple + params)
```
Can't this just be:
```
num_inputs = len(input_tuple)
_assertGradAndGradgradChecks(test_case, lambda *args: test_case._forward(module, args[num_inputs]), input_tuple + params)
```
?

glaringlee replied:
I will do a test. I think you are right!. I am so used to C++ placeholder thing......., will update this later.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4905,120 +4989,124 @@ def _do_test(self, test_case, module, input):
-            input_ip = deepcopy(input)
-            input_ip_clone = input_ip.clone()
+            input_ip = deepcopy(input_tuple)
+            input_ip_clone = tuple(t.clone() for t in input_ip)
             with freeze_rng_state():
-                output_ip = module_ip(input_ip_clone)
-            test_case.assertNotEqual(input_ip_clone._version, input_version)
+                output_ip = module_ip(*input_ip_clone)
+            input_ip_clone_versions = tuple(t._version for t in input_ip_clone)
+            test_case.assertNotEqual(input_ip_clone_versions, input_versions)
             test_case.assertEqual(output, output_ip)
             grad = output.data.clone().normal_()
-            input.grad.data.zero_()
+            for t in input_tuple:
+                t.grad.data.zero_()
             output.backward(grad)
             output_ip.backward(grad)
-            test_case.assertEqual(input.grad, input_ip.grad)
+            for t1, t2 in zip(input_tuple, input_ip):
+                test_case.assertEqual(t1.grad, t2.grad)
 
-        if isinstance(input, torch.LongTensor) and TEST_CUDA:
+        if any(isinstance(t, torch.LongTensor) for t in input_tuple) and TEST_CUDA:
             # check that cuda() moves module parameters to correct GPU device,
             # and that float() casts parameters correctly
-
-            input = input.cuda()
+            input_tuple = tuple(t.cuda() for t in input_tuple)
```
Can we just assert that `len(input_tuple) = 1` and use `input = input_tuple[0]` instead?

It doesn't quite make sense to me that an in-place version of a module should take in multiple tensors, because there is a confusion over "which of the multiple tensor arguments is actually being operated on in-place?". Furthermore Transformer has no in-place variant

glaringlee replied:
@zou3519 
I made this change to allow tuple inputs everywhere, not for Transformer only.
Do you mean we don't have the following cases?
1. Module.forward(_0, _1, _2...),  forward has multiple inputs and all of them are inplace var.
2. Module.forwar(_0, _1, _2...), forward has multiple inputs, and some of them are inplace var, eg (_0 and _2 are inplace)

The inplace check now can handle all above cases, I am not sure whether this is needed. If not, I will revert the changes to inplace_check and assert len(input_tuple) == 1.

zou3519 replied:
All of the inplace Modules that I am aware of (see [search](https://pytorch.org/docs/stable/search.html?q=inplace&check_keywords=yes&area=default), and by grepping common_nn.py for `check_inplace`) take a single Tensor as an input and modify it in-place. These consist of activation functions and dropout. I don't think we have any in-place modules that take in multiple tensors in the forward pass.

glaringlee replied:
I see. I will do `assert len(input_tuple) == 1` and leave a comment there.

Commit:
Update on "Support tuple inputs in NN Module test & add parity test for Transformer"


This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I enabled gradgrad check for TransformerEncoderLayer_relu test only since it is too slow. (10+ seconds for TransformerEncoderLayer, 50+ seconds for TransformerDecoderLayer, 10+ mins for multilayer Transformer). The underlying for these modules are all the same (Dropout, Linear and LayerNorm), so enabled 1 gradgrad test should be enough.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]

glaringlee commented:
@zou3519 Applied all the changes that we discussed previously.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,103 +4948,103 @@ def _do_test(self, test_case, module, input):
                     with torch.backends.cudnn.flags(enabled=False):
-                        module(input)
+                        module(*input_tuple)
                         for p in module.parameters():
                             test_case.assertIsInstance(p, torch.cuda.FloatTensor)
                             test_case.assertEqual(p.get_device(), 0)
 
                 if torch.cuda.device_count() >= 2:
                     # test cross-GPU transfer works
                     # to GPU1
-                    input = input.cuda(1)
+                    input_tuple = tuple(t.cuda(1) for t in input_tuple)
                     module.cuda(1)
                     with torch.cuda.device(1):
-                        module(input)
+                        module(*input_tuple)
                     for p in module.parameters():
                         test_case.assertIsInstance(p, torch.cuda.FloatTensor)
                         test_case.assertEqual(p.get_device(), 1)
 
                 if not self.skip_double:
                     # test double()
-                    input = input.double().cuda()
+                    input_tuple = tuple(t.double().cuda() for t in input_tuple)
                     module.double().cuda()
-                    module(input)
+                    module(*input_tuple)
                     for p in module.parameters():
                         test_case.assertIsInstance(p, torch.cuda.DoubleTensor)
                         test_case.assertEqual(p.get_device(), 0)
 
                 # test half()
-                input = input.half().cuda()
+                input_tuple = tuple(t.half().cuda() for t in input_tuple)
                 module.half().cuda()
-                module(input)
+                module(*input_tuple)
                 for p in module.parameters():
                     test_case.assertIsInstance(p, torch.cuda.HalfTensor)
                     test_case.assertEqual(p.get_device(), 0)
```
nit: There's a lot of repetition here in the device-swapping tests. It's a pre-existing problem but it would be nice if we could clean it up, since we're modifying the code anyways. It took me a while to understand what was being tested here.

glaringlee replied:
@zou3519 
Ah, yes, I understand this. I can consolidate this into one function and call it with different type and device id. 

When I wrote this part, I was debating on this. If written them into one function, then we loose the flexibility to change one of them if needed. And on the other hand, some of them test both assertIsInstance and assertEqual, some of them test assertIsInstance only. So if you consolidate them into one function, you will have some kind of flag indicate whether to test assertEqual, and the function call will be like foo(... true), foo(... false), etc. It might not as clear as what we have now.

zou3519 replied:
We could leave this for future work, it doesn't have to be done in this PR. It's a pre-existing problem and we don't have to block this PR on this.

> And on the other hand, some of them test both assertIsInstance and assertEqual, some of them test assertIsInstance only. So if you consolidate them into one function, you will have some kind of flag indicate whether to test assertEqual, and the function call will be like foo(... true), foo(... false), etc. It might not as clear as what we have now.

Actually, we could always do the isinstance and assertEqual checks. What those checks are checking is that (1) the Tensor has the correct dtype and (2) the tensor is on the correct device. For example,
- a lot of assertEqual(p.get_device()) calls are missing for cases where the input is on CPU. We can just assertEqual(p.get_device(), -1) for those.

zou3519 replied:
Oh, I see, you did this already. Thanks for improving the code!

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,103 +4948,103 @@ def _do_test(self, test_case, module, input):
             test_case.assertEqual(input.grad, input_ip.grad)
 
-        if isinstance(input, torch.LongTensor) and TEST_CUDA:
+        if any(isinstance(t, torch.LongTensor) for t in input_tuple) and TEST_CUDA:
```
In this PR, we've extended NN module tests to support inputs that are tuples. However, I don't think this conditional covers some cases. At a high level, what we're trying to test here is a cartesian product of the following conditions
- .cpu(), .cuda(), and .cuda(1) work on the model and inputs
- .float() and .double() and .half() work on the model and the floating-point inputs

Let's say there exists a module that takes in (LongTensor, LongTensor, DoubleTensor) inputs (it exists -- nn.EmbeddingBag. However, nn.EmbeddingBag doesn't actually get tested in this way because of the previous lack of tupled input support for NewModuleTest). Then in the current logic we wouldn't test the following cases:
- convert (LongTensor, LongTensor, DoubleTensor) to (LongTensor, LongTensor, FloatTensor)
- convert (LongTensor, LongTensor, DoubleTensor) to (LongTensor, LongTensor, HalfTensor)
- convert (LongTensor, LongTensor, DoubleTensor) to (cuda.LongTensor, cuda.LongTensor, cuda.FloatTensor)
- convert (LongTensor, LongTensor, DoubleTensor) to (cuda.LongTensor, cuda.LongTensor, cuda.HalfTensor)

glaringlee replied:
@zou3519  You are right, this is a place I was wondering whether I should use all instead of any. nn.EmbeddingBad should be supported once input_dict is supported. But if I change this to all we probably can support it now and still keep all previous cases running correctly.
Let me change this to all, agree?

zou3519 replied:
> So we actually don't care about whetehr we can convert for eg. from (LongTensor, LongTensor, DoubleTensor) to (cuda.LongTensor, cuda.LongTensor, cuda.FloatTensor), what we care about here is that whether LongTensor can be casted to FloatTensor correctly on cpu/gpu.

I disagree, I think the point of the test is to convert all floating-point inputs but not convert any integral tensor inputs.

Consider nn.EmbeddingBag. It takes in up to [three arguments in the forward pass](https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html).
```
model = nn.EmbeddingBag()
model(input: LongTensor, offsets: LongTensor, per_index_weights: Tensor)
```
If we changed `input` and/or `offsets` to FloatTensor, the model would break. This is because EmbeddingBag expects `input` and `offsets` to be a LongTensor but doesn't care what per_index_weights is. Most models that want an integral tensor require that tensor to be integral.

glaringlee replied:
@zou3519 You are partially right in my opinion. The code has two branches.
If input is LongTensor and TEST_CUDA. It actually test whether LongTensor can be convert to FloatTensor correctly by cuda. In this case, it does convert every tensor to Float.
The other branch is like what you said and I already changed the code in my previous PR, only convert non-Long Tensor.

glaringlee replied:
So for the first branch, now we only go into this branch if all input tensors are LongTensor.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4555,15 +4612,16 @@ def fw(input):
 
         # TODO: compare structure
-        if input.numel() != 0:
+        input_tuple = input if isinstance(input, tuple) else (input,)
+        if any(t.numel() != 0 for t in input_tuple):
```
What was the initial `if input.numel()` check guarding against? I'd like to make sure it is safe to do `if any(t.numel() != 0 for t in input_tuple)` here

glaringlee replied:
This is not a guard, just to save program running time. Previously we have only one input,  if that input is empty, then the jacobian output size will be 0 (jacobian output size is equal to the forward output's numel ), then no need to compare numerical jacobian and analytical jacobian. 
Now we have multiple tensors, so if any of them is not empty, then I assume we need to compare jacobian.

zou3519 replied:
Let's say the input was an empty tensor (so the input has size [0]). Then `input.grad` would have size `[0]`. The analytical jacobian would also have 0 numel (I think it get size [0, 0], but that doesn't really matter).

So in this case, `analytical_t = [ torch.zeros([0, 0]) ]`, `numerical_t = [ torch.zeros([0, 0]) ]`. 
If we tried to do `a.sub(n).abs().max()`, calling `.max()` on a tensor with no numel throws an exception!  So I think what we're guarding against here is calling .max() on a tensor with 0 numel().

I'm not 100% sure that the numerical jacobian actually comes out to be an empty tensor (it really depends on how we compute it...). But if the above is correct, then we should change the check to something like:
```
differences = []
for input, a, n in zip(input_tuple, analytical_t, numerical_t):
    if a.numel() == 0:
        continue   # TODO: compare structure (ensure analytic jacobian has correct shape)
    differences.append(a.add(n, alpha=-1).abs().max())
if differences):
    self.assertLessEqual(max(differences), PRECISION)
```

glaringlee replied:
@zou3519 
Ah, I did not check whether max will throw an error on empty tensor. I looked into the jacobian code, as long as one of input is not empty, the output size from jacobian will not be empty based on our current module. You are right, if max throw, the code will need to be like what you said. I will update the PR

zou3519 reviewed (commented):
Some more comments/questions from me. Most of the questions come from me not understanding the original code because it looks kinda weird. After these get resolved we should be good to go

Commit:
Update on "Support tuple inputs in NN Module test & add parity test for Transformer"


This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I enabled gradgrad check for TransformerEncoderLayer_relu test only since it is too slow. (10+ seconds for TransformerEncoderLayer, 50+ seconds for TransformerDecoderLayer, 10+ mins for multilayer Transformer). The underlying for these modules are all the same (Dropout, Linear and LayerNorm), so enabled 1 gradgrad test should be enough.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,90 @@ def _do_test(self, test_case, module, input):
             test_case.assertEqual(input.grad, input_ip.grad)
 
-        if isinstance(input, torch.LongTensor) and TEST_CUDA:
+        def test_module_parameters(dtype, device_id=None):
```
nit: `assert_module_parameters_are(tensor_type, device_id)` gets the idea across a bit better.

Also, the tensor type is not actually a "dtype" so we shouldn't call the argument "dtype" -- dtypes are things like `torch.float`, `torch.double`.

glaringlee replied:
sure

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,90 @@ def _do_test(self, test_case, module, input):
 
-        if isinstance(input, torch.LongTensor) and TEST_CUDA:
+        def test_module_parameters(dtype, device_id=None):
+            for p in module.parameters():
+                test_case.assertIsInstance(p, dtype)
+                if device_id is not None:
+                    test_case.assertEqual(p.get_device(), device_id)
+
+        if all(isinstance(t, torch.LongTensor) for t in input_tuple) and TEST_CUDA:
             # check that cuda() moves module parameters to correct GPU device,
             # and that float() casts parameters correctly
-
-            input = input.cuda()
+            input_tuple = tuple(t.cuda() for t in input_tuple)
             module.float().cuda()
-            module(input)
-            for p in module.parameters():
-                test_case.assertIsInstance(p, torch.cuda.FloatTensor)
-                test_case.assertEqual(p.get_device(), 0)
+            module(*input_tuple)
+            test_module_parameters(torch.cuda.FloatTensor, 0)
 
             if torch.cuda.device_count() > 1:
-                input = input.cuda(1)
+                input_tuple = tuple(t.cuda(1) for t in input_tuple)
                 module.cuda(1)
                 with torch.cuda.device(1):
-                    module(input)
-                for p in module.parameters():
-                    test_case.assertIsInstance(p, torch.cuda.FloatTensor)
-                    test_case.assertEqual(p.get_device(), 1)
+                    module(*input_tuple)
+                test_module_parameters(torch.cuda.FloatTensor, 1)
         else:
             # check that float()/double() casters work correctly
 
             # to float
-            if not isinstance(input, torch.LongTensor):
-                input = input.float()
+            input_tuple = tuple(t.float() if not isinstance(t, torch.LongTensor) else t for t in input_tuple)
```
flag

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,90 @@ def _do_test(self, test_case, module, input):
 
                 # to GPU0
-                input = input.float().cuda()
+                input_tuple = tuple(t.float().cuda() for t in input_tuple)
```
flag

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,90 @@ def _do_test(self, test_case, module, input):
 
                 # to CPU
-                input = input.cpu()
+                input_tuple = tuple(t.cpu() for t in input_tuple)
```
flag

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,90 @@ def _do_test(self, test_case, module, input):
                 if not self.skip_double:
                     # test double()
-                    input = input.double().cuda()
+                    input_tuple = tuple(t.double().cuda() for t in input_tuple)
```
possible flag

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,90 @@ def _do_test(self, test_case, module, input):
 
                 # test half()
-                input = input.half().cuda()
+                input_tuple = tuple(t.half().cuda() for t in input_tuple)
```
possible flag

Commit:
Update on "Support tuple inputs in NN Module test & add parity test for Transformer"


This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I enabled gradgrad check for TransformerEncoderLayer_relu test only since it is too slow. (10+ seconds for TransformerEncoderLayer, 50+ seconds for TransformerDecoderLayer, 10+ mins for multilayer Transformer). The underlying for these modules are all the same (Dropout, Linear and LayerNorm), so enabled 1 gradgrad test should be enough.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]

glaringlee commented:
@zou3519 
Applied all the changed we were agree on.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4881,106 +4948,93 @@ def _do_test(self, test_case, module, input):
 
                 # test half()
-                input = input.half().cuda()
+                input_tuple = tuple(
+                    t.half().cuda() if not isinstance(t, torch.LongTensor) else t.cuda() for t in input_tuple)
```
nit: We do this "cast if the input is not a LongTensor" a lot. Maybe we could abstract it away in the future.

glaringlee replied:
Just wondering how we abstract this in python? Use `to(dtype)` instead of .half, .double and .float? And then define a function which passed in the dtype to do the conversion?

zou3519 replied:
There are two possibilities here I think. The first is to define a function that takes in a lambda that has the cast:
```
def cast_if_fp(tensor, fp_cast):
    return cast(tensor) if torch.is_floating_point(tensor.dtype) else tensor
```
the second is as you said, to use `to(dtype)`. When we wrote NewModuleTest we didn't have `.to()` at our disposal.

glaringlee replied:
@zou3519 
ok, I get your point.

zou3519 replied:
The first approach (to use the cast) seems like it would be really wordy and not necessarily an improvement on the existing code. Either way, this isn't a blocker for the PR, just something I noticed

glaringlee replied:
Thanks for suggestions. I get your point.

Commit:
Update on "Support tuple inputs in NN Module test & add parity test for Transformer"


This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I enabled gradgrad check for TransformerEncoderLayer_relu test only since it is too slow. (10+ seconds for TransformerEncoderLayer, 50+ seconds for TransformerDecoderLayer, 10+ mins for multilayer Transformer). The underlying for these modules are all the same (Dropout, Linear and LayerNorm), so enabled 1 gradgrad test should be enough.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]

Commit:
Update on "Support tuple inputs in NN Module test & add parity test for Transformer"


This PR did two things:
1. Support tuple inputs for NNTestCase with NewModuleTest class. NewModuleTest now support passing multiple inputs in tuple through keyword input_fn. (see transformer test case). No need to separate src and target like criterion tests.
2. Add parity test for Transformer.
    TransformerEncoderLayer, TransformerDecoderLayer and Transformer are added into parity test.
    No need to add TransformerEncoder and TransformerDecoder since Transformer test have both of them tested and the underlying TransformerEncoderLayer and TransformerDecoderLayer are tested separately as well.
    I enabled gradgrad check for TransformerEncoderLayer_relu test only since it is too slow. (10+ seconds for TransformerEncoderLayer, 50+ seconds for TransformerDecoderLayer, 10+ mins for multilayer Transformer). The underlying for these modules are all the same (Dropout, Linear and LayerNorm), so enabled 1 gradgrad test should be enough.

Differential Revision: [D23750441](https://our.internmc.facebook.com/intern/diff/D23750441)

[ghstack-poisoned]
